Source/santad/DataLayer/SNTRuleTable.m

```diff
@@ -463,19 +463,44 @@ - (BOOL)addRules:(NSArray *)rules
 }
 
 - (BOOL)addedRulesShouldFlushDecisionCache:(NSArray *)rules {
-  // Check for non-plain-allowlist rules first before querying the database.
+  uint64_t nonAllowRuleCount = 0;
+
   for (SNTRule *rule in rules) {
-    if (rule.state != SNTRuleStateAllow) return YES;
+    if (rule.state != SNTRuleStateAllow) {
+      nonAllowRuleCount++;
+    }
   }
 
-  // If still here, then all rules in the array are allowlist rules.  So now we look for allowlist
-  // rules where there is a previously existing allowlist compiler rule for the same identifier.
-  // If so we find such a rule, then cache should be flushed.
+  // Just flush if we have a lot of block rules.
+  if (nonAllowRuleCount >= 1000) {
+    return YES;
+  }
+
+  // Check newly synced rules for any blocking rules. If any are found, check
+  // in the db to see if they already exist. If they're not found or were
+  // previously allow rules flush the cache.
+  //
+  // If all rules in the array are allowlist rules,  look for allowlist rules
+  // where there is a previously existing allowlist compiler rule for the same
+  // identifier.  If so we find such a rule, then cache should be flushed.
   __block BOOL flushDecisionCache = NO;
+
   [self inTransaction:^(FMDatabase *db, BOOL *rollback) {
     for (SNTRule *rule in rules) {
+      if (rule.state != SNTRuleStateAllow) {
+        if ([db longForQuery:
+                  @"SELECT COUNT(*) FROM rules WHERE identifier=? AND type=? AND state=? LIMIT 1",
+                  rule.identifier, @(rule.type), @(rule.state)] == 0) {
+          flushDecisionCache = YES;
+          break;
+        }
+        // Skip to check the next rule since we only want to check allows for
+        // compiler rules.
+        continue;
```

Comment:
Talked about this via video chat, but writing down for future us.

Technically  this check is correct. Since if we get an allow compiler rule, remove rule or block rule we need to invalidate the cache if there's potentially a corresponding allow rule.

The later check is unneeded as it just looks to see if there's a compiler rule in the database / cache that needs to be invalidated by a new allow rule.